diff --git i/Dockerfile w/Dockerfile
index 28a77e4..1c7e787 100644
--- i/Dockerfile
+++ w/Dockerfile
@@ -7,8 +7,6 @@
 # the extended buildx capabilities used in this file.
 # Make sure multiarch TARGETPLATFORM is available for interpolation
 # See: https://docs.docker.com/build/building/multi-platform/
-ARG TARGETPLATFORM=${TARGETPLATFORM}
-ARG BUILDPLATFORM=${BUILDPLATFORM}
 ARG BASE_REGISTRY="docker.io"
 
 # Ruby image to use for base image, change with [--build-arg RUBY_VERSION="3.4.x"]
diff --git i/streaming/Dockerfile w/streaming/Dockerfile
index 14f2d3c..0d8f8f3 100644
--- i/streaming/Dockerfile
+++ w/streaming/Dockerfile
@@ -4,8 +4,6 @@
 # the extended buildx capabilities used in this file.
 # Make sure multiarch TARGETPLATFORM is available for interpolation
 # See: https://docs.docker.com/build/building/multi-platform/
-ARG TARGETPLATFORM=${TARGETPLATFORM}
-ARG BUILDPLATFORM=${BUILDPLATFORM}
 ARG BASE_REGISTRY="docker.io"
 
 # Node version to use in base image, change with [--build-arg NODE_MAJOR_VERSION="20"]
