diff --git i/app/services/fetch_link_card_service.rb w/app/services/fetch_link_card_service.rb
index 4141fb43df..fd28f5efd2 100644
--- i/app/services/fetch_link_card_service.rb
+++ w/app/services/fetch_link_card_service.rb
@@ -48,7 +48,11 @@ class FetchLinkCardService < BaseService
     headers = {
       'Accept' => 'text/html',
       'Accept-Language' => "#{I18n.default_locale}, *;q=0.5",
-      'User-Agent' => "#{Mastodon::Version.user_agent} Bot",
+      'User-Agent' => if @url.match?(/\b(youtube\.com|youtu\.be)\b/)
+                        "Mozilla/5.0 (compatible; Discordbot/2.0; +https://discord.com)"
+                      else
+                        "#{Mastodon::Version.user_agent} Bot"
+                      end
     }
 
     @html = Request.new(:get, @url).add_headers(headers).perform do |res|
